src/structs/lepton_decoder.rs

```diff
@@ -282,77 +282,120 @@ fn parse_token<R: Read, const ALL_PRESENT: bool>(
     bool_reader: &mut VPXBoolReader<R>,
     image_data: &mut BlockBasedImage,
     context: &mut BlockContext,
-    num_non_zeros: &mut [NeighborSummary],
+    neighbor_summary_cache: &mut [NeighborSummary],
     qt: &QuantizationTables,
     pt: &ProbabilityTables,
     features: &EnabledFeatures,
 ) -> Result<()> {
     debug_assert!(pt.is_all_present() == ALL_PRESENT);
 
+    let neighbors =
+        context.get_neighbor_data::<ALL_PRESENT>(image_data, neighbor_summary_cache, pt);
+
+    let (output, ns) =
+        read_coefficient_block::<ALL_PRESENT, R>(pt, &neighbors, model, bool_reader, qt, features)?;
+
+    context.set_neighbor_summary_here(neighbor_summary_cache, ns);
+
+    image_data.append_block(output);
+
+    Ok(())
+}
+
+/// Reads the 8x8 coefficient block from the bit reader, taking into account the neighboring
+/// blocks, probability tables and model.
+///
+/// This function is designed to be independently callable without needing to know the context,
+/// image data, etc so it can be extensively unit tested.
+pub fn read_coefficient_block<const ALL_PRESENT: bool, R: Read>(
+    pt: &ProbabilityTables,
+    neighbor_data: &NeighborData,
+    model: &mut Model,
+    bool_reader: &mut VPXBoolReader<R>,
+    qt: &QuantizationTables,
+    features: &EnabledFeatures,
+) -> Result<(AlignedBlock, NeighborSummary)> {
     let model_per_color = model.get_per_color(pt);
 
+    // First we read the 49 inner coefficients
+
+    // figure out how many of these are non-zero, which is used both
+    // to terminate the loop early and as a predictor for the model
+    let predicted_num_non_zeros_7x7 =
+        pt.calc_non_zero_counts_context_7x7::<ALL_PRESENT>(neighbor_data);
```

Comment:
Hmm, the names are slightly misleading, as the number returned is from 0 to 25 - definitely not number of zeros, but just its predictor. What about `predicted_num_non_zeros_bin_7x7` and return bin? With corresponding changes in `read_non_zero_7x7_count` and in writer?

Reply:
made it num_non_zeros_7x7_context_bin

Reply:
Feel free to approve the review if you are ok with the changes